How many late days are you using for this assignment? 0

In [169]:
# Imports and pip installations (if needed)
%pip install pandas
import pandas as pd

Defaulting to user installation because normal site-packages is not writeable
You should consider upgrading via the '/Library/Developer/CommandLineTools/usr/bin/python3 -m pip install --upgrade pip' command.
Note: you may need to restart the kernel to use updated packages.


# Part 1: Load the dataset

In [170]:
# Load the given datasets
ckd_numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")
ckd_categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")
# Print the data
ckd_numerical
ckd_categorical

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:
Features: 
    Categorical:
    1. al, the amount of albunium found in a person's blood from a scale of 0-5
    2. su, the sugar intake from a scale of 0-5
    3. rbc, if the amount of red blood cells in the body is normal or abnormal
    Numerical: 
    1. unique_id, the unique identification number for each entry that determines which entry is which patient
    2. age, the age of the patient in years
    3. bp, the blood pressure of the patient, units in mm/Hg
Label: Target, whether a patient has or does not have chronic kidney disease

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [171]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicatesN = pd.DataFrame(ckd_numerical)[pd.DataFrame(ckd_numerical)['unique_id'].duplicated()]
print(len(duplicatesN))
# Drop these duplicate rows
ckd_numerical = pd.DataFrame(ckd_numerical).drop_duplicates(subset='unique_id')

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicatesC = pd.DataFrame(ckd_categorical)[pd.DataFrame(ckd_categorical)['unique_id'].duplicated()]
print(len(duplicatesC))
ckd_categorical = pd.DataFrame(ckd_categorical).drop_duplicates(subset='unique_id')

30
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No, all of the unique ids are NOT present in both datasets. This is because the categorical data set had 400 entries after removing duplicates, while the numerical data set had 390. Since the unique ids are indeed unique, there is no way for all of the unique ids to be present in both data sets. In the combined table, the ids that are in one dataset and not the other are excluded from the merged dataset.

In [172]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
ckd = pd.merge(pd.DataFrame(ckd_numerical), pd.DataFrame(ckd_categorical), on='unique_id')
df = pd.DataFrame(ckd)
#Print the combined dataset
ckd

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
385,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
388,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [173]:
# Calculate the percentage of rows that contain atleast one missing value
na_Percentage = df.isna().any(axis=1).sum()/df.shape[0]
# Print %
print(na_Percentage)
# Drop these rows from the dataset
df = df.dropna(axis=0)
# Print the Dataset
df

0.6076923076923076


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
13,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,...,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
379,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
381,462892,36.0,80.0,85.0,16.0,1.1,142.0,4.1,15.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.4: Sort the dataset according to the Labels

In [174]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
df = df.sort_values(by='Target')
df = df.reset_index(drop=True)
# Print the dataset
df

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,abnormal,present,notpresent,yes,yes,no,poor,no,yes,ckd
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [175]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
df = pd.get_dummies(ckd, columns=categorical_columns, dtype=int)

# Print the dataset
df

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_ckd\t,Target_notckd
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,0,1,0,1,0,1,0,0,0,1
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,0,1,0,1,0,1,0,0,0,1
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,0,1,0,1,0,1,0,0,0,1
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,...,0,0,1,0,1,1,0,1,0,0
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,...,0,1,0,1,0,1,0,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
385,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,...,0,0,1,0,1,1,0,1,0,0
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,0,1,0,0,1,1,0,1,0,0
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,0,1,0,1,0,1,0,0,0,1
388,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,0,1,0,1,0,0,1,1,0,0


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: I believe that just having a single column for the binary variable would be equivalent in the downstream modeling. However, if the categorical variable takes more than 2 values, then the downstream modeling would NOT be equivalent. This is because some condiments may be significant, and some condiment may not, however if you keep all of these in a single column you will not be able to determine which specific condiments are significant, you can only determine if condiments as a whole are significant.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [176]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
def remove_outliers_z_score(df, columns):
    # Create a mask as a Series of True
    mask = pd.Series(True, index=df.index)
    
    for column in columns:
        # Calculate Z-scores for the column
        mean = df[column].mean()
        std = df[column].std()
        
        # Calculate Z-scores
        z_scores = (df[column] - mean) / std
        
        # Update the mask considering a threshold for Z-scores (e.g., |Z| < 3)
        mask &= (z_scores.abs() < 3)
    
    # Filter the DataFrame based on the mask
    return df[mask]

# Remove outliers
df = remove_outliers_z_score(df, numerical_columns)
# Print the dataset
df

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_ckd\t,Target_notckd
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,0,1,0,1,0,1,0,0,0,1
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,0,1,0,1,0,1,0,0,0,1
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,0,1,0,1,0,1,0,0,0,1
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,...,0,0,1,0,1,1,0,1,0,0
5,992643,35.0,70.0,82.0,36.0,1.1,150.0,3.5,14.5,52.0,...,0,1,0,1,0,1,0,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
381,462892,36.0,80.0,85.0,16.0,1.1,142.0,4.1,15.6,44.0,...,0,1,0,1,0,1,0,0,0,1
383,838360,71.0,90.0,88.0,80.0,4.4,139.0,5.7,11.3,33.0,...,0,1,0,1,0,1,0,1,0,0
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,0,1,0,0,1,1,0,1,0,0
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,0,1,0,1,0,1,0,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [177]:
# Normalize the all Numerical Attributes in the dataset.
numeric_columns = df.select_dtypes(include=['number']).columns
df[numeric_columns] = (df[numeric_columns] - df[numeric_columns].min()) / (df[numeric_columns].max() - df[numeric_columns].min())
# Print the dataset
df

/var/folders/px/0vpsfw615yl0lhj4l6c9pp640000gn/T/ipykernel_47177/3665882024.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[numeric_columns] = (df[numeric_columns] - df[numeric_columns].min()) / (df[numeric_columns].max() - df[numeric_columns].min())


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_ckd\t,Target_notckd
0,0.113028,0.380952,0.6,0.093548,0.045455,0.006849,0.923077,0.196078,0.655738,0.736842,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
1,0.934673,0.440476,0.2,0.122581,0.075758,0.041096,0.846154,0.490196,1.000000,0.710526,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
2,0.356699,0.369048,0.2,0.132258,0.126263,0.027397,0.615385,0.313725,0.868852,0.894737,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
3,0.740083,0.761905,0.8,0.238710,0.580808,0.246575,0.641026,0.411765,0.524590,0.552632,...,0.0,0.0,1.0,0.0,1.0,1.0,0.0,1.0,NaN,0.0
5,0.995783,0.345238,0.4,0.038710,0.131313,0.047945,1.000000,0.196078,0.729508,0.947368,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
381,0.403045,0.357143,0.6,0.048387,0.030303,0.047945,0.794872,0.313725,0.819672,0.736842,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
383,0.823156,0.773810,0.8,0.058065,0.353535,0.273973,0.717949,0.627451,0.467213,0.447368,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,NaN,0.0
386,0.416816,0.690476,1.0,0.300000,0.222222,0.465753,0.743590,0.411765,0.188525,0.263158,...,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,NaN,0.0
387,0.073361,0.500000,0.2,0.135484,0.171717,0.054795,0.794872,0.470588,0.729508,0.736842,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The unique id is not appropriate for modeling nor predictions. This is because the unique id does not have any causation to having chronic kidney disease, the purpose for this column is to identify a data entry to a patient, which is not necessary for modeling and predictions. Additionally, the id makes sure each patient isn't recorded multiple times within the data, which after the duplicates are removed, the ids are not necessary.

In [178]:
#Remove that column
df = df.drop(columns='unique_id')
# Print the dataset
df

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_ckd\t,Target_notckd
0,0.380952,0.6,0.093548,0.045455,0.006849,0.923077,0.196078,0.655738,0.736842,0.271318,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
1,0.440476,0.2,0.122581,0.075758,0.041096,0.846154,0.490196,1.000000,0.710526,0.263566,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
2,0.369048,0.2,0.132258,0.126263,0.027397,0.615385,0.313725,0.868852,0.894737,0.131783,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
3,0.761905,0.8,0.238710,0.580808,0.246575,0.641026,0.411765,0.524590,0.552632,0.341085,...,0.0,0.0,1.0,0.0,1.0,1.0,0.0,1.0,NaN,0.0
5,0.345238,0.4,0.038710,0.131313,0.047945,1.000000,0.196078,0.729508,0.947368,0.434109,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
381,0.357143,0.6,0.048387,0.030303,0.047945,0.794872,0.313725,0.819672,0.736842,0.155039,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
383,0.773810,0.8,0.058065,0.353535,0.273973,0.717949,0.627451,0.467213,0.447368,0.534884,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,NaN,0.0
386,0.690476,1.0,0.300000,0.222222,0.465753,0.743590,0.411765,0.188525,0.263158,0.286822,...,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,NaN,0.0
387,0.500000,0.2,0.135484,0.171717,0.054795,0.794872,0.470588,0.729508,0.736842,0.434109,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [179]:
# Export the dataframe to a new csv file
df.to_csv("ckd_cleaned.csv")
# Export the dataframe to a new json file
df.to_json("ckd_cleaned.json")

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here

|    | Unnamed: 0 |    age    |   bp   |    bgr    |     bu     |    sc    |    sod    |   pot    |   hemo   |    pcv   | ... | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_ckd | Target_notckd |
|----|-------------|-----------|--------|-----------|------------|----------|-----------|----------|----------|----------|-----|---------|------------|------------|-------|--------|--------|---------|------------|------------|----------------|
|  0 |      0      | 0.380952  |  0.6   |  0.093548 |  0.045455  | 0.006849 |  0.923077 | 0.196078 | 0.655738 | 0.736842 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |
|  1 |      1      | 0.440476  |  0.2   |  0.122581 |  0.075758  | 0.041096 |  0.846154 | 0.490196 | 1.000000 | 0.710526 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |
|  2 |      2      | 0.369048  |  0.2   |  0.132258 |  0.126263  | 0.027397 |  0.615385 | 0.313725 | 0.868852 | 0.894737 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |
|  3 |      3      | 0.761905  |  0.8   |  0.238710 |  0.580808  | 0.246575 |  0.641026 | 0.411765 | 0.524590 | 0.552632 | ... | 0.0     | 0.0        | 1.0        |  0.0  |  1.0   |  1.0   |  0.0    |      NaN   |     0.0   |                |
|  4 |      5      | 0.345238  |  0.4   |  0.038710 |  0.131313  | 0.047945 |  1.000000 | 0.196078 | 0.729508 | 0.947368 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |
|  5 |      7      | 0.690476  |  0.2   |  0.116129 |  0.085859  | 0.020548 |  1.000000 | 0.156863 | 0.721311 | 0.684211 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |
|  6 |      8      | 0.321429  |  0.2   |  0.193548 |  0.156566  | 0.034247 |  0.769231 | 0.372549 | 0.811475 | 0.947368 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |
|  7 |     10      | 0.666667  |  0.8   |  0.319355 |  0.191919  | 0.136986 |  0.692308 | 0.078431 | 0.639344 | 0.815789 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     0.0   |                |
|  8 |     17      | 0.619048  |  0.6   |  0.196774 |  0.040404  | 0.047945 |  0.769231 | 0.196078 | 0.836066 | 0.973684 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |
|  9 |     18      | 0.202381  |  0.2   |  0.080645 |  0.070707  | 0.027397 |  0.871795 | 0.490196 | 0.770492 | 0.947368 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |
| 10 |     19      | 0.511905  |  0.6   |  0.167742 |  0.111111  | 0.054795 |  0.717949 | 0.274510 | 0.934426 | 0.657895 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |
| 11 |     20      | 0.488095  |  0.2   |  0.125806 |  0.075758  | 0.047945 |  0.769231 | 0.431373 | 0.836066 | 0.657895 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |
| 12 |     22      | 0.500000  |  0.6   |  0.016129 |  0.060606  | 0.027397 |  0.666667 | 0.490196 | 0.918033 | 0.921053 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |
| 13 |     23      | 0.321429  |  0.6   |  0.061290 |  0.045455  | 0.047945 |  0.846154 | 0.490196 | 0.770492 | 0.631579 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |
| 14 |     24      | 0.321429  |  0.2   |  0.032258 |  0.075758  | 0.034247 |  0.897436 | 0.196078 | 0.696721 | 0.842105 | ... | 0.0     | 1.0        | 0.0        |  1.0  |  0.0   |  1.0   |  0.0    |      NaN   |     1.0   |                |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

I could mitigate data conversion hallucinations from LLM conversions by being very clear when prompting. Additionally, I can prompt multiple times and see if the output is any different. I can also split the data into smaller and easier sections in order for the large language model to limit its hallucinations.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [180]:
# Code to rename all the columns in the dataset
new_column_names = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}
df.rename(columns=new_column_names, inplace=True)
# Print the dataset
df

,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,packed cell volume,wbcc,...,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_ckd\t,Target_notckd
0,0.380952,0.6,0.093548,0.045455,0.006849,0.923077,0.196078,0.655738,0.736842,0.271318,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
1,0.440476,0.2,0.122581,0.075758,0.041096,0.846154,0.490196,1.000000,0.710526,0.263566,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
2,0.369048,0.2,0.132258,0.126263,0.027397,0.615385,0.313725,0.868852,0.894737,0.131783,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
3,0.761905,0.8,0.238710,0.580808,0.246575,0.641026,0.411765,0.524590,0.552632,0.341085,...,0.0,0.0,1.0,0.0,1.0,1.0,0.0,1.0,NaN,0.0
5,0.345238,0.4,0.038710,0.131313,0.047945,1.000000,0.196078,0.729508,0.947368,0.434109,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
381,0.357143,0.6,0.048387,0.030303,0.047945,0.794872,0.313725,0.819672,0.736842,0.155039,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0
383,0.773810,0.8,0.058065,0.353535,0.273973,0.717949,0.627451,0.467213,0.447368,0.534884,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,1.0,NaN,0.0
386,0.690476,1.0,0.300000,0.222222,0.465753,0.743590,0.411765,0.188525,0.263158,0.286822,...,0.0,1.0,0.0,0.0,1.0,1.0,0.0,1.0,NaN,0.0
387,0.500000,0.2,0.135484,0.171717,0.054795,0.794872,0.470588,0.729508,0.736842,0.434109,...,0.0,1.0,0.0,1.0,0.0,1.0,0.0,0.0,NaN,1.0


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [182]:
# Converted SQL to Pandas code
result = df.groupby('Target_ckd').size().reset_index(name='count')

*Describe what the above code does here*
The code above groups the entries by their Target value, and records the number of entries in each group. Since there are two different values in Target, the data entries are grouped by not having chronic kidney disease and by having chronic kidney disease.